docs/propagation.md

```diff
@@ -210,3 +217,58 @@ along F0, propagate `["c"]` along F1, and propagate nothing along F2.
 T0  | "a", **"b"** | **"c"**  | "f" |
 T1  | "a", "b"     | "c", "d" | "g" |
 T2  | **"a", "b"** | "c", "e" |     |
+
+### Data flow ops
+
+The above propagation step description applies to most ops. However, there are
+cases where a sharding rule is not appropriate. For those cases, Shardy defines
+_data flow _ ops.
+
+A data flow edge of some op X defines a bridge between a set of *sources* and a
+set of *targets*, such that all sources and targets should be sharded in the
+same way. Examples of such ops are `stablehlo::OptimizationBarrierOp`,
+`stablehlo::WhileOp`, `stablehlo::CaseOp` and also [`sdy::ManualComputationOp`](./sdy_dialect#sdymanual_computation_sdymanualcomputationop). Ultimately, any op that implements [ShardableDataFlowOpInterface](sdy_op_interfaces#shardabledataflowopinterface_shardabledataflowopinterface)
+is considered a data flow op.
+
+An op can have multiple data flow edges that are orthogonal to one another. For
+example:
+
+```mlir
+    y_0, ..., y_n = while (x_0, ..., x_n)
+                    ((pred_arg_0,... , pred_arg_n) { ... })
+                    ((body_arg_0,..., body_arg_n) {
+                    ...
+                    return return_value_0, ..., return_value_n
+                    })
+```
+
+This while op has `n` data flow edges: the i-th data flow edges is between
+sources `x_i`, `return_value_i` and targets `y_i`, `pred_arg_i`,
+`body_arg_i`.
+
+Shardy will propagate shardings between all sources and targets of a
+`sdy.data_flow_edge` as if it was a regular op with the sources as operands
```

Comment:
This is the first time we are mentioning this sdy.data_flow_edge op, so it's out of context. Can we avoid referring to the op and just say between all sources and targets of a data flow edge?